libraries/Microsoft.Bot.Connector/OAuthApiClient.cs

```diff
@@ -0,0 +1,350 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.Rest;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Connector
+{
+    /// <summary>
+    /// Service client to handle requests to the botframework api service.
+    /// </summary>
+    public class OAuthClient : ServiceClient<OAuthClient>
+    {
+        private readonly ConnectorClient client;
+        private readonly string uri;
+
+
+        public OAuthClient(ConnectorClient client, string uri)
+        {
+            Uri uriResult;
+            if (!(Uri.TryCreate(uri, UriKind.Absolute, out uriResult) && uriResult.Scheme == Uri.UriSchemeHttps))
+                throw new InvalidOperationException("Please supply a valid https uri");
+            this.client = client ?? throw new ArgumentNullException(nameof(client));
+            this.uri = uri;
+        }
+
+        /// <summary>
+        /// Get User Token for given user and connection.
+        /// </summary>
+        /// <param name="userId"></param>
+        /// <param name="connectionName"></param>
+        /// <param name="magicCode"></param>
+        /// <param name="customHeaders"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<TokenResponse> GetUserTokenAsync(string userId, string connectionName, string magicCode, Dictionary<string, List<string>> customHeaders = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (string.IsNullOrEmpty(userId))
+            {
+                throw new ArgumentNullException(nameof(userId));
+            }
+            if (string.IsNullOrEmpty(connectionName))
+            {
+                throw new ArgumentNullException(nameof(connectionName));
+            }
+
+            // Tracing
+            bool shouldTrace = ServiceClientTracing.IsEnabled;
+            string invocationId = null;
+            if (shouldTrace)
+            {
+                invocationId = ServiceClientTracing.NextInvocationId.ToString();
+                Dictionary<string, object> tracingParameters = new Dictionary<string, object>();
+                tracingParameters.Add("userId", userId);
+                tracingParameters.Add("connectionName", connectionName);
+                tracingParameters.Add("magicCode", magicCode);
+                tracingParameters.Add("cancellationToken", cancellationToken);
+                ServiceClientTracing.Enter(invocationId, this, "GetUserTokenAsync", tracingParameters);
+            }
+            // Construct URL
+            var tokenUrl = new Uri(new Uri(uri + (uri.EndsWith("/") ? "" : "/")), "api/usertoken/GetToken?userId={userId}&connectionName={connectionName}{magicCodeParam}").ToString();
```

Comment:
Please move all scary magic strings into constants with detailed comments on them.

Reply:
i want to keep this logic as such for now, since it's in sync wtih v3 sdk and node as well (v3 and v4) . i will update all sdk's to use a proper uribuilder for this logic in another review.